### PR TITLE
windows: add multiple timestamps

### PR DESCRIPTION
**Pull request by gdams**
fixes: https://github.com/AdoptOpenJDK/openjdk-build/issues/1963

test job: https://ci.adoptopenjdk.net/job/build-scripts/job/release/job/standalone_create_installer_windows/117/console

### CODE REVIEW

**Comment by karianna** [comment]
@gdams - did it work?

File: wix/Build.OpenJDK_generic.cmd [hunk 1]
```diff
@@ -307,6 +337,3 @@ SET WIN_SDK_FULL_VERSION=
 SET WIN_SDK_MAJOR_VERSION=
 
 EXIT /b 0
-
-:FAILED
-EXIT /b 2
```
Comment (douph1) [inline wix/Build.OpenJDK_generic.cmd:312]: Why remove this ? 
There is GOTO FAILED on top of the script

**Comment by douph1** [comment]
proposed fix and some improvement here https://github.com/gdams/openjdk-installer/pull/1

**Comment by karianna** [comment]
> proposed fix and some improvement here [gdams#1](https://github.com/gdams/openjdk-installer/pull/1)

CC @gdams

**Comment by douph1** [comment]
another timeout seen yesterday

**Comment by karianna** [comment]
@gdams - I think this is a release stability PR that we _could_ land with careful review - is this something you can complete and test this week?

**Comment by gdams** [comment]
@douph1 are you able to finish the PR? I think we still had some failures when we last tested this...

**Comment by douph1** [comment]
See all my comments in my original  PR https://github.com/gdams/openjdk-installer/pull/1

**Comment by douph1** [comment]
- retry logic seems abusive to me .. 300 and not gradual.
- list of timestamp server must be updated

File: wix/Build.OpenJDK_generic.cmd (hunk 1 above)
Comment (douph1) [reply wix/Build.OpenJDK_generic.cmd:312]: ?

Comment (gdams) [reply wix/Build.OpenJDK_generic.cmd]: added it back

**Comment by gdams** [comment]
okay this is updated and ready to go, In principle, this could go in before the CPU releases to avoid the need for a rebuild if the signing job fails due to timeout

CC @karianna

File: wix/serverTimestamp.config [hunk 2]
```diff
@@ -0,0 +1,5 @@
+; comment with semicolon
+; blank line are ingnored
+; script use only first column
+http://timestamp.verisign.com/scripts/timstamp.dll
```
Comment (karianna) [inline wix/serverTimestamp.config:4]: Is that the right UR? missing an `e`?

Comment (gdams) [reply wix/serverTimestamp.config]: fixed

**Comment by douph1** [comment]
We have a little pb here .. it seems a lot of url are different for sha1/sha256 sign
and we do both : >REM Dual-Signing with SHA-1/SHA-256 requires Win 8.1 SDK or later.

So we must add a param to url or manage two different list of url ..
Also some timestamp service seems to use same url for both sha1/sha256 .. but this is not clear to me


- A lot of references spell it timstamp.dll and not timestamp.dll and it is a dead service :
https://knowledge.digicert.com/alerts/migration-of-legacy-verisign-and-symantec-time-stamping-services

It must be replaced with SHA1/SHA256  http://timestamp.digicert.com?alg=sha256


Also comodoca don't use /authenticode anymore

https://support.comodo.com/index.php?/Knowledgebase/Article/View/68/0/time-stamping-server

https://gist.github.com/Manouchehri/fd754e402d98430243455713efada710#gistcomment-3342213
>But sectigo (comodoca) is only the certificate of sha1's certificate chain expired on May 31st, 2020. It's not a surprise. This certificate has long been prescribed.

Also for sha256 : http://timestamp.comodoca.com?td=sha256


So maybe we must probably append sha1/sha256 to url used with signtools when we required one or the other algo

We probably can use more timestamp server list as this maintained list : https://gist.github.com/Manouchehri/fd754e402d98430243455713efada710
(if we lower the retry delay)

https://stackoverflow.com/questions/9714798/http-timestamp-verisign-com-scripts-timstamp-dll-not-available

Sorry to not help so much as I'm not on Windows and can't test which timestamp url/service works ..

File: wix/serverTimestamp.config (hunk 2 above)
Comment (douph1) [reply wix/serverTimestamp.config]: @gdams  have you tried both with sha1 and sha256 ?

**Comment by karianna** [comment]
> We have a little pb here .. it seems a lot of url are different for sha1/sha256 sign
> and we do both : >REM Dual-Signing with SHA-1/SHA-256 requires Win 8.1 SDK or later.
> 
> So we must add a param to url or manage two different list of url ..
> Also some timestamp service seems to use same url for both sha1/sha256 .. but this is not clear to me
> 
> * A lot of references spell it timstamp.dll and not timestamp.dll and it is a dead service :
>   https://knowledge.digicert.com/alerts/migration-of-legacy-verisign-and-symantec-time-stamping-services
> 
> It must be replaced with SHA1/SHA256 http://timestamp.digicert.com?alg=sha256
> 
> Also comodoca don't use /authenticode anymore
> 
> https://support.comodo.com/index.php?/Knowledgebase/Article/View/68/0/time-stamping-server
> 
> https://gist.github.com/Manouchehri/fd754e402d98430243455713efada710#gistcomment-3342213
> 
> > But sectigo (comodoca) is only the certificate of sha1's certificate chain expired on May 31st, 2020. It's not a surprise. This certificate has long been prescribed.
> 
> Also for sha256 : http://timestamp.comodoca.com?td=sha256
> 
> So maybe we must probably append sha1/sha256 to url used with signtools when we required one or the other algo
> 
> We probably can use more timestamp server list as this maintained list : https://gist.github.com/Manouchehri/fd754e402d98430243455713efada710
> (if we lower the retry delay)
> 
> https://stackoverflow.com/questions/9714798/http-timestamp-verisign-com-scripts-timstamp-dll-not-available
> 
> Sorry to not help so much as I'm not on Windows and can't test which timestamp url/service works ..

CC - @gdams JIC

**Comment by douph1** [comment]
I have try to signing localy a exe and see that signing with" -fd sha56" like we did override the first sha1 signature.
So double signing doesn't work at this time .. but it must be possible with two certs
dualsign

**Comment by douph1** [comment]
and dual signing msi are visibly particulary painfull 
https://social.msdn.microsoft.com/Forums/ie/en-US/d4b70ecd-a883-4289-8047-cc9cde28b492/sha1-sha256-dualsigning-for-msi?forum=windowssecurity

https://support.ksoftware.net/support/solutions/articles/217399-how-do-i-dual-sign-a-file-
>MSI and other file formats typically do not support dual signing.

If all our msi are already signed only with sha256 can't we stick with that ?
It will simplify script and timestamp server list

**Comment by gdams** [comment]
@douph1 after talking to some MSFT folks, SHA1 signing has been deprecated for some time. They recommend that we just use SHA256 so I'll update the PR>

**Comment by karianna** [comment]
@douph1 can you review again?

**Comment by douph1** [comment]
I will asap

**Comment by douph1** [comment]
I have done the last modifications here https://github.com/gdams/openjdk-installer/pull/2
please review and merge @gdams then I will approve this one

**Comment by gdams** [comment]
@douph1 merged! thanks

**Review by douph1 — APPROVED** [review]

**Pull request merged**